spec/datadog/appsec/processor_spec.rb

```diff
@@ -0,0 +1,227 @@
+# typed: ignore
+
+require 'datadog/appsec/spec_helper'
+require 'datadog/appsec/processor'
+
+RSpec.describe Datadog::AppSec::Processor do
+  before do
+    skip 'disabled for Java' if RUBY_PLATFORM.include?('java')
+
+    logger = double(Datadog::Core::Logger)
+    allow(logger).to receive(:debug?).and_return true
+    allow(logger).to receive(:debug)
+    allow(logger).to receive(:info)
+    allow(logger).to receive(:warn)
+    allow(logger).to receive(:error)
+
+    allow(Datadog).to receive(:logger).and_return(logger)
+  end
+
+  context 'self' do
+    it 'detects if the WAF is unavailable' do
+      hide_const('Datadog::AppSec::WAF')
+
+      expect(described_class.libddwaf_required?).to be false
+    end
+
+    it 'detects if the WAF is available' do
+      stub_const('Datadog::AppSec::WAF', Module.new)
+
+      expect(described_class.libddwaf_required?).to be true
+    end
+
+    it 'reports via return of libddwaf loading failure' do
+      allow(Object).to receive(:require).with('libddwaf').and_raise(LoadError)
+
+      expect(described_class.require_libddwaf).to be false
+    end
+
+    it 'reports via return of libddwaf loading success (first require)' do
+      allow(Object).to receive(:require).with('libddwaf').and_return(true)
+
+      expect(described_class.require_libddwaf).to be true
+    end
+
+    it 'reports via return of libddwaf loading success (second require)' do
+      allow(Object).to receive(:require).with('libddwaf').and_return(false)
+
+      expect(described_class.require_libddwaf).to be true
+    end
+  end
+
+  describe '#load_libddwaf' do
+    context 'when LoadError is raised' do
+      before do
+        allow(Object).to receive(:require).with('libddwaf').and_raise(LoadError)
+      end
+
+      it { expect(described_class.new.send(:load_libddwaf)).to be false }
+    end
+
+    context 'when loaded but missing mandatory const' do
+      before do
+        allow(Object).to receive(:require).with('libddwaf').and_return(true)
+        hide_const('Datadog::AppSec::WAF')
+      end
+
+      it { expect(described_class.new.send(:load_libddwaf)).to be false }
+    end
+
+    context 'when loaded successfully' do
+      before do
+        allow(Object).to receive(:require).with('libddwaf').and_return(true)
+        stub_const('Datadog::AppSec::WAF', Module.new)
+      end
+
+      it { expect(described_class.new.send(:load_libddwaf)).to be true }
+    end
+  end
```

Comment:
These specs are duplicating the specs for `require_libddwaf` and `libddwaf_required?` above.

May I suggest perhaps turning `require_libddwaf` and `libddwaf_required?` into private methods and then just testing them through `load_libddwaf`? Or alternatively, just go with a whitebox testing approach and test that `load_libddwaf` delegates to the other two methods, rather than repeating their tests.

Reply:
The coverage is a bit different. I had a choice at implementing either a) as two unit-test blocks, complete with internals stubbing, b) a single functional one with all cases, or unit-testing one with all internal cases + functional testing one with fewer (but higher level) cases. I chose the latter, thinking it would be more obvious in case of failure.